### DISCUSSION TITLE
Typst pass a measurement variable and type conversion

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#8051] [Q&A] [answered]
### Description

I have created this minimal example from [letter](https://github.com/quarto-ext/typst-templates/tree/main/letter) template.

In this template, I want the vertical spacing to be passed in a metadata variable. So I have added `gap`.

**template.typ**

```
#let letter(
  recipient: none,
  gap: none,
  body
) = {
  set page(paper: "us-letter", margin: (top: 2cm))
  recipient
  v(gap)
  body
}
```

**show.typ**

```
#show: letter.with(
$if(recipient)$
  recipient: [$recipient$],
$endif$
$if(gap)$
  gap: [$gap$],
$endif$
)
```

**letter.qmd**

```
---
recipient: Mr. John Doe
gap: 2cm
format:
  letter-typst: default
---

Dear Joe,

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque tempus lorem quam, nec eleifend risus rhoncus sed. Fusce urna quam, placerat eu nisi eget, consequat tempor mauris. Nullam maximus est sit amet blandit venenatis. Curabitur vestibulum, nulla eget bibendum blandit, est nisi convallis massa, vel tincidunt orci mi dictum magna. Duis rhoncus neque vitae ipsum imperdiet pulvinar. 

Best,
Jane Smith
Regional Director
```

I get this error:

```
[typst]: Compiling letter.typ to letter.pdf...error: expected relative length or fraction, found content
    ┌─ letter.typ:159:4
    │
159 │   v(gap)
    │     ^^^
```

I couldn't really find any information on type conversion in typst [type documentation](https://typst.app/docs/reference/foundations/type/). Alternatively, it would be nice to read in the parameter as a number and then create the unit.
`gap = 1 + "cm"`, but that didn't work:

```
[typst]: Compiling letter.typ to letter.pdf...error: cannot add integer and string
    ┌─ letter.typ:161:7
    │
161 │   type(1+"cm")

[typst]: Compiling letter.typ to letter.pdf...error: cannot join content with type
    ┌─ letter.typ:161:2
    │
161 │   type("1"+"cm")
```

`quarto 1.4.533`

**Comment by cderv** [comment, marked as answer]
This is a typst issue: 
```
[typst]: Compiling letter.typ to letter.pdf...error: expected relative length or fraction, found content
```

You are providing 
````
  gap: [$gap$],
````
so this will be 
````
  gap: [5cm],
````
which [content](https://typst.app/docs/reference/foundations/content/)  type, and you don't want that. You want a [fraction](https://typst.app/docs/reference/layout/fraction/) or [relative length](https://typst.app/docs/reference/layout/relative/). 

So based on the doc, you probably only need 
````
$if(gap)$
  gap: $gap$,
$endif$
````

and this in your YAML

````yaml
gap: 5cm
````

> Alternatively, it would be nice to read in the parameter as a number and then create the unit.

Pandoc's templating is just string substition in template doc. So if you do 
````
$if(gap)$
  gap: $gap$cm,
$endif$
````

and  in yaml 
````
gap: 5
````

it would be as if you write in `.typ`
````
gap: 5cm
````


Overall, you need to learn more about 
* Pandoc's template (https://pandoc.org/MANUAL.html#templates)
* Typst feature and type (https://typst.app/docs/)

BTW, typst website has a really cool feature which is an online app that will create the doc https://typst.app/ 

You can easily check that the `.typ` you are generating (add `keep-typ: true` in YAML to get it) will work, and you can see errors, how to fix them and try of few syntax. Great way to learn. 

When you have a working `.typ` document, you can update your pandoc's template for Quarto accordingly

Hope it helps

**Reply by royfrancis** [reply]
That works! Thank you! I incorrectly assumed the `[ ]` in show was for optional arguments. I can't remember where I got that idea from.
